### PR TITLE
remove buildkit as dependency from the CLI (integrate github.com/moby/buildkit/util/appcontext)

### PR DESCRIPTION
**Pull request by thaJeztah**
- relates to https://github.com/docker/cli/pull/4455#issuecomment-1644032209
- relates to https://github.com/moby/patternmatcher/pull/1

This copies the github.com/moby/buildkit/util/appcontext
package as an internal package. The appcontext package from
BuildKit was the only remaining dependency on BuildKit, and
while we may need some of its functionality, the implementation
is not correct for how it's used in docker/cli (so would need
a rewrite).

Moving a copy of the code into the docker/cli (but as internal
package to prevent others from depending on it) is a first step
in that process, and removes the circular dependency between
BuildKit and the CLi.

We are only using these:

    tree vendor/github.com/moby/buildkit
    vendor/github.com/moby/buildkit
    ├── AUTHORS
    ├── LICENSE
    └── util
        └── appcontext
            ├── appcontext.go
            ├── appcontext_unix.go
            ├── appcontext_windows.go
            └── register.go

    3 directories, 6 files

Before this:

    go mod graph | grep ' github.com/docker/cli'
    github.com/moby/buildkit@v0.11.6 github.com/docker/cli@v23.0.0-rc.1+incompatible

After this:

    go mod graph | grep ' github.com/docker/cli'
    # (nothing)

But indirect dependencies and circular dependency on moby/moby -> buildkit -> moby/moby and docker/cli can cause "hell".

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: cmd/docker/appcontext/appcontext.go [hunk 1]
```diff
@@ -9,8 +9,10 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-var appContextCache context.Context
-var appContextOnce sync.Once
+var (
+	appContextCache context.Context
+	appContextOnce  sync.Once
+)
```
Comment (thaJeztah) [inline cmd/docker/appcontext/appcontext.go:15]: oh! that was `gofumpt` (which I have enabled in my IDE), well 🤷

**Comment by thaJeztah** [comment]
Quick blurb, because we discussed this in one of the maintainers calls with @tonistiigi @neersighted (and others, but they may recall the exact details);

- the current use of `appcontext` in the CLI is not correct (also see https://github.com/docker/cli/issues/4402)
- ... but we may want to keep _parts_ of the behavior
- and.. this is where my memory fails me 🙈 😊 because I don't recall the exact parts we should _keep_ and which parts to _remove_
- I _think_ the "CTRL-C x3 to _force_ stop" was something that was desirable
- But ... I forgot the other details 🙈

**Comment by thaJeztah** [comment]
@tonistiigi @neersighted PTAL; this keeps the status-quo on the functionality and does not fix the issues that were discussed about appcontext, but does remove the BuildKit dependency, which I think may be worth considering.

I updated the PR to move the appcontext package to `internal` (at least until we fixed the issues that were discussed).

**Review by neersighted — APPROVED** [review]

**Comment by thaJeztah** [comment]
Let me bring this one in; we can use https://github.com/docker/cli/issues/4402 as tracking issue for the remaining work

**Pull request merged**